src/platform/Linux/ConnectivityManagerImpl.cpp

```diff
@@ -66,5 +83,196 @@ void ConnectivityManagerImpl::_OnPlatformEvent(const ChipDeviceEvent * event)
 #endif
 }
 
+#if CHIP_DEVICE_CONFIG_ENABLE_WPA
+
+uint16_t ConnectivityManagerImpl::mConnectivityFlag;
+struct GDBusWpaSupplicant ConnectivityManagerImpl::mWpaSupplicant;
+
+bool ConnectivityManagerImpl::_HaveIPv4InternetConnectivity(void)
+{
+    return ((mConnectivityFlag & kFlag_HaveIPv4InternetConnectivity) != 0);
+}
+
+bool ConnectivityManagerImpl::_HaveIPv6InternetConnectivity(void)
+{
+    return ((mConnectivityFlag & kFlag_HaveIPv6InternetConnectivity) != 0);
+}
+
+bool ConnectivityManagerImpl::_IsWiFiStationConnected(void)
+{
+    bool ret            = false;
+    const gchar * state = NULL;
+
+    if (mWpaSupplicant.state != GDBusWpaSupplicant::WPA_INTERFACE_CONNECTED)
+    {
+        ChipLogProgress(DeviceLayer, "wpa_supplicant: _IsWiFiStationConnected: interface not connected");
+        return false;
+    }
+
+    state = wpa_fi_w1_wpa_supplicant1_interface_get_state(mWpaSupplicant.iface);
+    ChipLogProgress(DeviceLayer, "wpa_supplicant: wpa_fi_w1_wpa_supplicant1_interface_get_state: %s", state);
+
+    if (g_strcmp0(state, "completed") == 0)
+    {
+        ChipLogProgress(DeviceLayer, "wpa_supplicant: WiFi already connected");
+        SetFlag(mConnectivityFlag, kFlag_HaveIPv4InternetConnectivity, true);
+        SetFlag(mConnectivityFlag, kFlag_HaveIPv6InternetConnectivity, true);
+        ret = true;
+    }
+
+    return ret;
+}
+
+bool ConnectivityManagerImpl::_CanStartWiFiScan()
+{
+    return mWpaSupplicant.state == GDBusWpaSupplicant::WPA_INTERFACE_CONNECTED &&
+        mWpaSupplicant.scanState == GDBusWpaSupplicant::WIFI_SCANNING_IDLE;
+}
+
+void ConnectivityManagerImpl::_OnWpaInterfaceProxyReady(GObject * source_object, GAsyncResult * res, gpointer user_data)
+{
+    GError * err                            = NULL;
+    WpaFiW1Wpa_supplicant1Interface * iface = wpa_fi_w1_wpa_supplicant1_interface_proxy_new_for_bus_finish(res, &err);
+
+    if (mWpaSupplicant.iface)
+    {
+        g_object_unref(mWpaSupplicant.iface);
+        mWpaSupplicant.iface = NULL;
+    }
+
+    if (iface != NULL && err == NULL)
+    {
+        mWpaSupplicant.iface = iface;
+        mWpaSupplicant.state = GDBusWpaSupplicant::WPA_INTERFACE_CONNECTED;
+        ChipLogProgress(DeviceLayer, "wpa_supplicant: connected to wpa_supplicant interface proxy: %p", iface);
+    }
+    else
+    {
+        if (err != NULL)
```

Comment:
Suggest removing this branch.

Can it really happen that |err| is null here?

If it can't happen,
  assert(err);

If it can happen,
  ChipLogProgress(DeviceLayer, "wpa_supplicant: failed to create wpa_supplicant1 interface proxy %s: %s",
mWpaSupplicant.interfacePath, err ? err->message : "unknown error");

Reply:
Ack, in theory, err should never be NULL if iface is NULL, but lets not assume that.